### PR TITLE
Reject requests to create packages having the name 'default'

### LINKED ISSUE #1785: make package name "default" reserved

### PR DESCRIPTION
**Pull request by mdeuser**
Updates to existing packages named "default" are still permitted.

Fixes #1785

### CODE REVIEW

**Comment by mdeuser** [comment]
pg3/1333

File: core/controller/src/main/scala/whisk/core/controller/Packages.scala [hunk 1]
```diff
@@ -76,25 +76,30 @@ trait WhiskPackagesApi extends WhiskCollectionAPI with ReferencedEntities {
    */
   override def create(user: Identity, entityName: FullyQualifiedEntityName)(implicit transid: TransactionId) = {
     parameter('overwrite ? false) { overwrite =>
-      entity(as[WhiskPackagePut]) { content =>
-        val request = content.resolve(entityName.namespace)
+      if (!overwrite && entityName.name.toString == "default") {
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Packages.scala]: iirc there’s an asString method which is preferred.

File: core/controller/src/main/scala/whisk/core/controller/Packages.scala [hunk 2]
```diff
@@ -76,25 +76,30 @@ trait WhiskPackagesApi extends WhiskCollectionAPI with ReferencedEntities {
...
+        logging.info(this, "attempt to create a package named 'default'")
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Packages.scala]: We can probably do away with this log message or use debug level

File: tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala [hunk 3]
```diff
@@ -370,6 +370,20 @@ class PackagesApiTests extends ControllerTestCommon with WhiskPackagesApi {
     }
   }
 
+  it should "reject create package when package name is 'default'" in {
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala]: Is there a test for update of an existing package named default?

File: tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala [hunk 4]
```diff
@@ -370,6 +370,20 @@ class PackagesApiTests extends ControllerTestCommon with WhiskPackagesApi {
...
+    implicit val tid = transid()
+    def defname() = MakeName.next("default")
+    val provider = WhiskPackage(namespace, defname(), None)
+    Put(s"$collectionPath/${provider.name}") ~> Route.seal(routes(creds)) ~> check {
+      status should be(BadRequest)
+      val response = responseAs[String]
+      println(s"Default failure response: '${response}'")
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala]: Printing not necessary.

File: core/controller/src/main/scala/whisk/core/controller/Packages.scala (hunk 1 above)
Comment (rabbah) [reply core/controller/src/main/scala/whisk/core/controller/Packages.scala]: I’d move the string default to a const on the whisk package Singleton.

Comment (mdeuser) [reply core/controller/src/main/scala/whisk/core/controller/Packages.scala]: :+1:  at the same time, i'll make it an array of reserved names that can be expanded as needed

File: tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala (hunk 3 above)
Comment (mdeuser) [reply tests/src/test/scala/whisk/core/controller/test/PackagesApiTests.scala]: not currently; i'll add one

**Comment by rabbah** [comment]
Hmm I think this update logic is incorrect - it would allow me to bypass the reserved names by just doing a PUT force=true. I believe the intention was to allow this _only_ if a package already exists with that name and otherwise to reject it. @mdeuser can you confirm and open a defect if you agree.

**Comment by mdeuser** [comment]
@rabbah - yep.  a package update will allow the "default" name for the package.  need to add a check to confirm whether or not the package exists.  hopefully something simple exists :-)

**Comment by dubeejw** [comment]
Looks like we can still create a package with the default name if the package is created by `wsk action update`.

**Comment by dubeejw** [comment]
@mdeuser, seems like if we won't let a package to be created named as `default` we shouldn't let packages with the `default` name be updated either?

**Comment by mdeuser** [comment]
@dubeejw - yep.  @rabbah mentioned the same issue.  I opened https://github.com/apache/incubator-openwhisk/pull/2964 to address it

**Comment by mdeuser** [comment]
@dubeejw - if a package exists with a reserved name, i believe the intent is not to block updates to that package.

**Pull request merged**